llvm/lib/Transforms/Scalar/SeparateConstOffsetFromGEP.cpp

```diff
@@ -1017,6 +1016,51 @@ bool SeparateConstOffsetFromGEP::reorderGEP(GetElementPtrInst *GEP,
       IsChainInBounds &= KnownPtrGEPIdx.isNonNegative();
     }
   }
+  TypeSize GEPSize = DL->getTypeSizeInBits(GEP->getSourceElementType());
+  TypeSize PtrGEPSize = DL->getTypeSizeInBits(PtrGEP->getSourceElementType());
```

Comment:
You're working on the wrong type sizes here. You need to use the alloc sizes.

But more generally, maybe I don't get what this code is doing, but can't you also swap the GEP source element types when swapping the indices? That's how you usually reassociate GEPs. Why do you need/want to rewrite them?

Reply:
> maybe I don't get what this code is doing

I think you probably do .. We're looking at GEPs that have ptrs produced by another GEP. If the inner GEP has a constant offset and the outer GEP doesn't, we reorganize so the outer GEP has the constant offset. Previously, this was only supported for GEPs with the same alloc type; this patch extends it somewhat by allowing for some differences in alloc type and doing the necessary accounting in the index. 

> Why do you need/want to rewrite them?

We're not requiring one use on the nested GEP, so we can't write it in place. Not sure if it makes sense to try to write the outer GEP in place if we're changing it's pointer?

Reply:
I'm not suggesting that you perform an in-place rewrite, but rather than you preserve the original element types of the GEPs. If you have `gep t2, (gep t1, p, i1), i2`, you should be able to rewrite it to `gep t1 (gep t2, p, i2), i1` without changing types or index values.

Reply:
I was trying to preserve the pointee type for users of the outer GEP, but I don't have a compelling argument as to why we would want to do this.